### ISSUE TITLE
Confusion regarding the "learnable token" in the iterative camera prediction

### ISSUE THREAD

**Issue by NIRVANALAN** [facebookresearch/vggt#51]
Hi, thanks for sharing this great work.

In the (iterative) camera prediction pipeline, the CameraHead inits the prediction from a "learnable" token, as below:

https://github.com/facebookresearch/vggt/blob/6830466c427cd77c877ae2d4705c13fc684a890f/vggt/heads/camera_head.py#L68t

However, since you detach the gradient in each step, the gradient from the final pose loss will not flow back to the torch.zeros() tokens right (since you use the last step prediction for supervision)? Thus, it is not a learnable token?

https://github.com/facebookresearch/vggt/blob/f1bc88b9f040420a197b4c91bb1d49f146ad698f/vggt/heads/camera_head.py#L126

Thanks and looking forward to your reply.

**Comment by jytime** [comment]
Hi it is a learnable token. However, we only expect self.empty_pose_tokens be updated for the first iteration, as below: 

```
            if pred_pose_enc is None:
                module_input = self.embed_pose(self.empty_pose_tokens.expand(B, S, -1))
            else:
                # Detach the previous prediction to avoid backprop through time.
                pred_pose_enc = pred_pose_enc.detach()
                module_input = self.embed_pose(pred_pose_enc)
```

We hold the idea that, we should somehow tell the model that, it is the first prediction and no prior can rely on. One possible solution is just inputing all zero or identity as the init. However, sometimes all zero or identity are also meaningful (e.g., some cameras can just be identity). Therefore, we use a learnable token to indicate this state, and that's why we call it empty pose token.

**Comment by NIRVANALAN** [comment]
I see, thanks for your clarification. 

My concern lies in how you supervise the training, since according to this line of code
https://github.com/facebookresearch/vggt/blob/6830466c427cd77c877ae2d4705c13fc684a890f/vggt/models/vggt.py#L70

vggt only returns the final iteration camera prediction result. So this is only for inference right, during training you supervise all the intermediate camera prediction?

Thanks.

**Comment by jytime** [comment]
yes during training we supervise all intermediate results similar to

https://github.com/facebookresearch/co-tracker/blob/82e02e8029753ad4ef13cf06be7f4fc5facdda4d/cotracker/models/core/cotracker/losses.py#L14

**Comment by NIRVANALAN** [comment]
I see, thanks for your prompt reply. By the way, for your depth loss supervision, would you share the (or refer to) a similar implementation? E.g., the SIGLogLoss + GradL1 loss in https://github.com/isl-org/ZoeDepth/blob/edb6daf45458569e24f50250ef1ed08c015f17a7/zoedepth/trainers/loss.py#L42

Thanks again.

**Comment by jytime** [comment]
This https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/model_components/losses.py looks similar but I did not check the details of their implementation

**Comment by NIRVANALAN** [comment]
Thanks for your prompt reply, and your reference code works well!

By the way, may I ask whether the DINO encoder is jointly trained or frozen during the vgg-t model training? Full-attention on 24 views is memory intensive and I wonder how you addressed it.

Finally, the Alternative Attention design was also used in our previous work [LN3Diff](https://github.com/NIRVANALAN/LN3Diff?tab=readme-ov-file) in the  3D VAE training stage (the self-plane, cross-plane attention). Sincerely appreciate if you can refer/cite our work in your future work/extensions!

![Image](https://github.com/user-attachments/assets/2c70a387-d1c1-4cbd-b4aa-85f1539e3da9)

Thanks again for open-sourcing this fabulous project ;)

**Comment by jytime** [comment]
Hi, 

The DINO part is not frozen. For full attention over 24 views or more, you need flash attention2, bf16, and torch checkpointing. 

Thanks i will have a look at LN3Diff

**Comment by NIRVANALAN** [comment]
Thanks Jianyuan!
